move/sources/canvas/token.move

```diff
@@ -388,39 +333,47 @@ module addr::canvas_token {
             let r = vector::pop_back(&mut rs);
             let g = vector::pop_back(&mut gs);
             let b = vector::pop_back(&mut bs);
-            draw_one(caller_addr, canvas, x, y, r, g, b);
+            draw_one(canvas_, x, y, r, g, b);
             i = i + 1;
         };
     }
 
     /// Draw a single pixel to the canvas. We consider the top left corner 0,0.
     public fun draw_one(
-        caller_addr: address,
-        canvas: Object<Canvas>,
-        x: u64,
-        y: u64,
+        canvas_: &mut Canvas,
+        x: u16,
+        y: u16,
         r: u8,
         g: u8,
         b: u8,
-    ) acquires Canvas {
-        let cost = determine_cost(canvas, x, y);
-
-        let canvas_ = borrow_global_mut<Canvas>(object::object_address(&canvas));
-
+    ) {
         // Confirm the coordinates are not out of bounds.
         assert!(x < canvas_.config.width, error::invalid_argument(E_COORDINATE_OUT_OF_BOUNDS));
         assert!(y < canvas_.config.height, error::invalid_argument(E_COORDINATE_OUT_OF_BOUNDS));
 
-        // If there is a cost, take the PNT and burn it.
-        if (cost > 0) {
-            paint_fungible_asset::burn(caller_addr, cost);
-        };
-
         // Write the pixel.
-        let color = Color { r, g, b };
+        let color = combine_rgb(r, g, b);
         let pixel = Pixel { color, drawn_at_s: now_seconds() };
         let index = y * canvas_.config.width + x;
-        smart_table::upsert(&mut canvas_.pixels, index, pixel);
+        smart_table::upsert(&mut canvas_.pixels, (index as u32), pixel);
+    }
+
+    /// Combine RGB values into a single u8 using bitwise operations
+    fun combine_rgb(r: u8, g: u8, b: u8): u8 {
+        assert!(r <= 255 && g <= 255 && b <= 255, error::invalid_argument(E_INVALID_RGB));
+        let r_u32 = (r as u32);
+        let g_u32 = (g as u32);
+        let b_u32 = (b as u32);
+        let combined = (r_u32 << 16) | (g_u32 << 8) | b_u32;
+        (combined & 0xFF as u8)
+    }
+
+    // Extract the R, G, and B components using bitwise operations and masks
+    fun extract_rgb(combined_color: u8): (u8, u8, u8) {
```

Comment:
we probably don't need this `extract_rgb` in the contract, but we are going to need it in the processor @banool

Reply:
also... we're taking 3 u8s, and we're somehow storing them inside of a single u8? Maybe I am missing something, but I don't see how that is possible